snuba/web/query.py

```diff
@@ -103,7 +104,9 @@ def _run_query_pipeline(
     for processor in dataset.get_query_processors():
         processor.process_query(request.query, request.settings)
 
-    storage_query_plan = dataset.get_query_plan_builder().build_plan(request)
+    query_plan_builder = dataset.get_query_plan_builder()
+    storage_query_plan = query_plan_builder.build_plan(request)
+    reader = query_plan_builder.get_storage(request).get_cluster().get_reader()
```

Comment:
I don't think this is going to work for joined queries. We need to assume multiple storages can be used in the same query. I think there are 2 issues here:
1) easier to solve: I am not comfortable in leaking the storage class at this point from the query plan builder. The plan buidler returns a query plan, and that is supposed to know how to run the query (it can run a query across multiple storages which are at that point hidden from this code). You can let the plan builder provide a cluster to the query plan and get the cluster from the queryplan. See an example of how it could work (I need to wrap that PR) here https://github.com/getsentry/snuba/pull/830/files#diff-87784c0acdd074c79dca3bec3bb2d403R16.
2) I would not even leak the reader here. Why don't you just let the QueryPlanBuilder provide the cluster or the StorageSet to the QueryPlanExecutionStrategy that will pass the cluster to the QueryRunner  which will resolve the reader and do the rest ? I am suggesting this because it should make the information flow easier to understand. We initialize the query runner only with the data the QueryPlan cannot know about (like the timer). The rest (that is the outcomes of the decisions made by QueryPlanBuilder) flows directly from QueryPlanBuilder to QueryExecutionStrategy. But I do not have a strong opinion on this last point. As long as we do not depend on a single storage and we do not leak the storage class here I am ok.

Reply:
still applies

Reply:
Thanks, I've been looking for a better solution for this. I changed the builder to pass the cluster to the query plan execution strategy.